### ISSUE TITLE
Explore UX for three-way merge

### ISSUE THREAD

**Issue by daviddossett** [microsoft/vscode#146091]
We're exploring what it could look like to enable a three-way merge workflow in VS Code. Here are some initial ideas to get the ball rolling 👍

## Prototype

Here's a brief walkthrough where I set up a simple merge conflict and resolve it with a three-way merge editor:

https://user-images.githubusercontent.com/25163139/160215589-6a275452-5f92-4341-9a00-1c57e345f120.mp4

## Main views

Here are some quick screenshots from the prototype above:

### Entry point

I've been exploring the idea of having multiple entry points from a merge conflict state. One is via the existing codelens actions ("Preview Merge")...

<img width="2159" alt="entry-point-codelens" src="https://user-images.githubusercontent.com/25163139/160215002-a7aced62-3a47-46f1-bdd2-40b58e5f888d.png">

..and another via the SCM where there could be an icon button (icon here is just a placeholder) on the file that has conflicts:

<img width="594" alt="entry-point-scm" src="https://user-images.githubusercontent.com/25163139/160215010-95fb3d4f-b84d-48d7-bb4b-32a7c943395a.png">

### Preview merge

Here's what the merge editor preview could look like. Here it features a mixed layout with the branches on top and the resulting code below. The branch editors are read-only while the result editor is editable. 

<img width="2150" alt="default-state" src="https://user-images.githubusercontent.com/25163139/160215096-d1a858ac-d36d-4a7e-8ecb-d910a9ccf6bd.png">

### With selections

Here are two examples where selections have been made. One with a selections from different conflicts on each branch:

<img width="2154" alt="with-selections" src="https://user-images.githubusercontent.com/25163139/160215165-516ca50d-07ff-4360-aa11-87ecab1e3ccb.png">

...and another showing the same changes being picked on both branches:

<img width="2158" alt="accept-both-changes" src="https://user-images.githubusercontent.com/25163139/160215182-9edc132e-6a33-4d7d-9a69-0c4b709bbcde.png">

Feedback and ideas welcome! 

cc @misolori @albertelo @lszomoru @rebornix

**Comment by Snailedlt** [comment]
Would be nice to see what branch incoming and current branches are.
I always struggle to remember which one is incoming and whoch one is current. 

Maybe something like showing the branchname in parenthesis next to current/incoming, or showing the branchname on hover would be a good option?

**Comment by hawkticehurst** [comment]
Feedback:

**Entry point:** 
Why not have both? I personally question if I would have noticed either of the entry points that you demonstrated (or at least I think it would've taken me a while to notice), so having multiple opportunities to discover the new feature feels better to me than aligning on one of them. 

**Preview merge:**
I love the mixed layout view, but more generally think that offering multiple layouts is the way to go.

One potential nit is the "inconsistency" between panel order in the horizontal/vertical modes––I was expecting the results tab to be in the middle of the horizontal layout based on you demoing the vertical layout first. With that said, I could kind of see how you were maybe trying to align the mixed layout and horizontal layouts by having the results on the bottom?

I don't think I have a strong opinion about what the horizontal/vertical layout order should be (since again I can see myself almost exclusively using the mixed layout) but probably worth discussing this a bit more/getting community feedback on. Or maybe even giving the option to let people adjust the vertical/horizontal layout order to suit their preferences?

**With selections:**

I actually really like the usage of checkboxes as a mechanism for selecting lines of code (iirc this is not done elsewhere in VS Code right?). I personally think it's a really clear UI/UX model.

To offer another suggestion that came to mind while watching this: a visually cleaner (but potentially less intuitive) model could be to remove the checkboxes and simply click on the highlighted lines of code you want to choose from the incoming and current panels (since they are read-only). I almost imagine it as a sort of toggle button-esque experience since you've already designed a checked (higher opacity colors/text) and unchecked (lower opacity colors/text) state for the lines of code themselves.

I still think I prefer the usage of checkboxes though.

**Misc:**

_Setting to make this a default_
I love what you've presented so much that I can easily see myself wanting this to be the default way I manage merge conflicts in VS Code (if it's implemented), so I just want to toss in a formal request that a setting to make three-way merge the default view when managing merge conflicts be added as well. Please and thank you 🙏😊

_Icon buttons are no longer icon buttons?_
More of a concern for future webview UI toolkit work, but I just wanted to make sure that there was an intentional decision to break the conventions of the icon button to include text in the top right corner of each panel? 

I personally really like how it looks (compared to using a secondary button for example) and think the addition of text is a really important and useful way of distinguishing between the actions of each panel. But again just want to make sure this an intentional shift in how the icon button style is used?

**Comment by hawkticehurst** [comment]
> Would be nice to see what branch incoming and current branches are.

@Snailedlt I think that's already been included in the design 😊

<img width="2150" alt="three-way-merge" src="https://user-images.githubusercontent.com/39639992/160218865-6afb96ff-607a-4704-8956-a7989f35ee13.png">

**Comment by bpasero** [comment]
@daviddossett really like it and seems to pretty much match Visual Studio in terms of UX which I felt was already a great UI when I had used it.

Can we also make sure to get VS behaviour of auto-picking every change that can be merged without conflicts, such as in this example where a change is selected for getting applied automatically since it does not result in conflicts:

![image](https://user-images.githubusercontent.com/900690/160228861-7f82c25d-94a5-48bf-a786-bf29094ddef3.png)

I think another entry point could be a prominent action in the editor toolbar to open the merge editor. Nevertheless, I think all the entry points you suggest work for me. We probably also want a setting to let the user choose to always open the merge editor by default when opening a file with merge markers. I personally would not want to always having to click on the entry point to get into this mode, I would want it always. 

Have we thought about how to represent a change in a line? Throwing in some ideas from others:

![image](https://user-images.githubusercontent.com/900690/160229025-cad66b51-243b-4153-a231-1a2fc0ef00a1.png)

![image](https://user-images.githubusercontent.com/900690/160229027-1ef3a7b3-1d21-4449-b9d6-53a37c3cbce0.png)

The reason I bring this up is https://github.com/microsoft/vscode/issues/25887, where people ask for a way to take changes from left or right even when just comparing 2 files. I think if we solve this for the merge editor, the same solution could be used for diff editors too?

**Comment by Snailedlt** [comment]
> > Would be nice to see what branch incoming and current branches are.
> 
> @Snailedlt I think that's already been included in the design 😊
> 
> <img width="2150" alt="three-way-merge" src="https://user-images.githubusercontent.com/39639992/160218865-6afb96ff-607a-4704-8956-a7989f35ee13.png">

Ohh, you're right!

Well in that case it all looks great!

**Comment by flumacs** [comment]
> I think another entry point could be a prominent action in the editor toolbar to open the merge editor. Nevertheless, I think all the entry points you suggest work for me. We probably also want a setting to let the user choose to always open the merge editor by default when opening a file with merge markers. I personally would not want to always having to click on the entry point to get into this mode, I would want it always.


similarly to what @bpasero said, would it be possible to open this way all the conflicting files - by default - with a simple command within the CLI ?

**Comment by bpasero** [comment]
👍 , I think the requirements for the command line are captured in https://github.com/microsoft/vscode/issues/5770 and we want to make sure that is supported.

**Comment by Zerotask** [comment]
It would be great to see how many remaining conflicts exists (and how many got resolved). 

### Example:
![a](https://user-images.githubusercontent.com/20150243/160378803-3ed9544a-002d-4d92-afa6-28f4bbe55e65.png)

**Comment by isidorn** [comment]
Love the direction @daviddossett 👏 
Some feedback I already shared, just to repeat here for transparency:
* You are suggesting that we introduce a new editor title area. I think this works for this use case, not sure if we could get away without this?

fyi for editor @alexdima @hediet, for scm thoughts @lszomoru @joaomoreno

**Comment by misolori** [comment]
Overall, really digging the design and direction! I do think we'll need to have a separate exploration around improving the codelenses for the conflicts as the multiple actions can be hard to comprehend quickly.

> I just wanted to make sure that there was an intentional decision to break the conventions of the icon button to include text in the top right corner of each panel?

+1 to this thought, we also already use a "Merge" button action for settings sync so we should make both of these button styles consistent either way we go. I'm more inclined to go with button style to avoid introducing another style (though it is used in some areas like the notebook toolbar).

<img width="785" alt="CleanShot 2022-03-28 at 11 20 48@2x" src="https://user-images.githubusercontent.com/35271042/160461740-6c14888d-f5fd-47e5-a324-0b6b3a5dee84.png">

**Comment by Snailedlt** [comment]
> +1 to this thought, we also already use a "Merge" button action for settings sync so we should make both of these button styles consistent either way we go. I'm more inclined to go with button style to avoid introducing another style (though it is used in some areas like the notebook toolbar).
> 
> <img width="785" alt="CleanShot 2022-03-28 at 11 20 48@2x" src="https://user-images.githubusercontent.com/35271042/160461740-6c14888d-f5fd-47e5-a324-0b6b3a5dee84.png">
> 

I'm more for the option shown in the preview, where the "take all incoming" prompt is on the topbar. I feel it's a more natural place for it to be. It's not as intrusive ad the bright blue "pop up" buttons, which takes focus away from, and covers up some of the code. 
The pop-up button style is more useful in cases where you need to react to something unusual, or something that happens then and there. Since it covers up the code, I like to think of it as a "notification.

Meanwhile the topbar button is more useful when it's a recurring choice, that happens every time (which is the case for merge conflicts, where you always want to get the option to accept all changes).


TLDR: Topbar button is better in this case since it's less intrusive, while the bright blue pop-up button is better for notification type prompts.

**Comment by hediet** [comment]
One problem I often have with merge conflicts is to understand the changes on the right and on the left individually before trying to combine them.
For simple renames this is easy to figure out, but for reorderings of members or complex refactorings it is not so obvious.

I think it could be helpful to show more details of the commits between merge-base and the respective branch (maybe even filtering out those commits that did not touch the file).

Maybe the most recent commit message + author could be shown next to the branch name. More commits could be indicated by three dots.
Clicking on such a commit could show the diff between that commit and the merge-base.

I wonder if we could innovate here a bit by trying to figure out what the text edit is (A rename of a word? A reordering of some lines?) to help the user to apply the effect of one side to the other.

**Comment by bpasero** [comment]
Such a concept exists in Eclipse as "Structure Compare". I couldn't find a good screenshot, but the gist is:

![image](https://user-images.githubusercontent.com/900690/160619854-3c0f79de-e628-471e-b521-4d778cf45407.png)

**Comment by daviddossett** [comment]
Thanks for all the feedback so far! Responding to a handful of comments before jumping into some variations:

@hawkticehurst 

> To offer another suggestion that came to mind while watching this: a visually cleaner (but potentially less intuitive) model could be to remove the checkboxes and simply click on the highlighted lines of code you want to choose from the incoming and current panels (since they are read-only).

A similar suggestion came up to use the code lens actions above each change to avoid a new pattern. Each change could have an action that flips the state. A bit clunky perhaps, but one less thing to learn.

> Icon buttons are no longer icon buttons?

This is actually using the icon button component from our design toolkit. We have a variant for an icon w/ label. @misolori are these used anywhere in VS Code? I swear I'd seen them but I can't find any from a quick dig around. In any case I'll explore some other variations here—see comments below re: settings sync.

<img width="582" alt="CleanShot 2022-03-30 at 16 03 20@2x" src="https://user-images.githubusercontent.com/25163139/160944927-875abe79-44a5-4480-9a96-b28dd9895b7b.png">

@bpasero 

> Can we also make sure to get VS behaviour of auto-picking every change that can be merged without conflicts, such as in this example where a change is selected for getting applied automatically since it does not result in conflicts:

I didn't actually realize they included changes that didn't result in conflicts. Good to know. It makes this more of a general diff merge workflow rather than dealing specifically with conflicts 👍 

>I think another entry point could be a prominent action in the editor toolbar to open the merge editor. Nevertheless, I think all the entry points you suggest work for me. We probably also want a setting to let the user choose to always open the merge editor by default when opening a file with merge markers. I personally would not want to always having to click on the entry point to get into this mode, I would want it always.

Good idea. @hawkticehurst had the same feedback above. I agree—open to suggestions for how this setting could look if anyone has strong opinions.

> Have we thought about how to represent a change in a line? 

Not yet but I'll take a look at these examples + others.

@misolori 

>  I do think we'll need to have a separate exploration around improving the codelenses for the conflicts as the multiple actions can be hard to comprehend quickly.

I'll get a new issue going for this.

> +1 to this thought, we also already use a "Merge" button action for settings sync so we should make both of these button styles consistent either way we go. I'm more inclined to go with button style to avoid introducing another style (though it is used in some areas like the notebook toolbar

This is on my radar to explore as another option this week 👍

**Comment by misolori** [comment]
@daviddossett I believe this is only used in Notebooks for the toolbar:

<img width="550" src="https://user-images.githubusercontent.com/35271042/160973724-d13db9aa-0c18-48ee-ba1d-86b1df1e8446.png">

<img width="541" src="https://user-images.githubusercontent.com/35271042/160973665-043e2838-2b78-45aa-aca2-4d79864665f0.png">

**Comment by daviddossett** [comment]
👋 Following up with some variations based on some of the feedback/suggestions above. I'm sure there will be more to come later as there are other interesting ideas above that I haven't addressed yet.

## Alternate "Accept all" button styles and placement

### Titlebar with small primary buttons

One option would be to use a small primary button as found in the extensions list view. While this idea of a new toolbar for each pane in the editor is still new, there would at least be some connection to other buttons in VS Code compared to the initial idea of the icon w/ label:

<img width="1268" alt="pane-button-small-primary" src="https://user-images.githubusercontent.com/25163139/161324212-a2cc4e86-58a6-4971-861c-eb87e7cbb525.png">

### Fixed footer with primary buttons

Another idea explores @misolori's suggestion to be consistent with the settings sync merge could yield something like this:

<img width="1268" alt="pane-button-fixed-footer" src="https://user-images.githubusercontent.com/25163139/161324583-b15f09fb-6271-4d93-af1b-c207bb3ef181.png">

### Toolbar actions

One other idea would be to simply use the toolbar for all three actions. Less discoverable, but no net new UI here:

<img width="610" alt="accept-toolbar-actions" src="https://user-images.githubusercontent.com/25163139/161325004-6701e5b7-d03f-425a-9740-e02ffefa4455.png">

## Using CodeLens actions instead of checkboxes

This examples reuses the UI and interaction people are used to in our existing merge conflicts UI for picking changes. It's slightly different in that the actions act as toggles in this case.

https://user-images.githubusercontent.com/25163139/161324839-a2e59abc-7268-4d72-ae8f-8a6cf4cd14d6.mp4

## Removing the title bar for each pane

Following up on @isidorn's suggestion, we could try _not_ using a title bar for each pane and instead rely on other cues to indicate which pane is the incoming, current, and result. I think these are somewhat difficult to understand, but it's worth thinking more about how we could avoid introducing a new split view w/ title bar concept.

Both options lack context re: branch name which makes it especially hard to know what's happening. Regardless, some jumping off points could look like:

### Using the hints from the buttons

<img width="1268" alt="no-titlebar-checkboxes-buttons" src="https://user-images.githubusercontent.com/25163139/161325541-542c675b-eb56-4ca6-be6e-c5649e25fd26.png">

### Using more specific CodeLens text

<img width="1268" alt="no-titlebar-codelens" src="https://user-images.githubusercontent.com/25163139/161325572-ac3055a7-bba0-41fe-bd86-82b629040a05.png">

**Comment by hediet** [comment]
Personally, I'd prefer to not add too much noise to the source code. The colors are already quite noisy and I already find it difficult to not get lost in the changes. If there is a good design without inline code-lenses, I'd go for that.

**Comment by hediet** [comment]
What happens if the user manually edits the result view in a way that incoming changes can no longer be accepted/rejected automatically?
(E.g. when the entire result is deleted)

**Comment by Snailedlt** [comment]
Would it be possible to have the option to show the three windows side-by-side similar to what Jetbrains IDE's do?

That makes it easier to see what line the changes affect without looking up and down at the line numbers. It also gives us the opportunity to distinguish the two changes from eachother by saying "left" and "right" instead of the commit-hash, which is a bit easier to understand for beginners.
I also think the way "accept changes", and "decline changes" is handled with arrow and x symbols instead of with text is quite intuitive.

Picture for reference:
![merge-conflict](https://user-images.githubusercontent.com/43886029/161520038-a4e5802d-0c8c-4d0f-bd3c-46f7e82a1e1c.png)

**Comment by isidorn** [comment]
@daviddossett great explorations, thanks a lot!
I think this convinced me that we can not get away without introducing one title area per merge editor.

As for the way to accept changes - I agree with @hediet we should not add noise to the source code. And based on that moto I prefer the checkboxes in the glyph area on the left side.

**Comment by hediet** [comment]
> Precedence: the debug toolbar is floating around.

I'm not a huge fan of the floating debug toolbar to be honest - it always covers something. I don't know if we should use that as precedence.

**Comment by byehack** [comment]
> > I think this convinced me that we can not get away without introducing one title area per merge editor.
> 
> I really dislike having a title area, cause itself is a container in a hiearchy of containers with title areas. We could have an alternative affordance which, like the huge buttons in bottom right, floats in that corner of the editor but is a toolbar which contains small action icons. Precedence: the debug toolbar is floating around.

Just using title bar for `Result` editor.
`[Accept | Take All] Incoming/Current` can go in `Result` title bar. (I think it has enough space in Mixed Layout mode)

![image](https://user-images.githubusercontent.com/26390254/161584082-c98f2bdf-3ae0-4c1e-8e61-6066d1368e0a.png)

I also agree with code lense in `Result` editor. so we can jump to Incoming/Current editor by clicking on code lenses in result editor.

**Comment by daviddossett** [comment]
> We also utilize the tab title in regular diffs as well, as an additional area to surface information

I had thought about something like that. One challenge would be not knowing that the bottom pane is the result until you have sort of seen the effect of taking/discarding changes. Open to other ideas here though.

<img width="1448" alt="CleanShot 2022-04-04 at 20 58 10@2x" src="https://user-images.githubusercontent.com/25163139/161676160-69fc927f-0c9a-435e-a94b-d65a408705e3.png">

**Comment by daviddossett** [comment]
> Would it be possible to have the option to show the three windows side-by-side similar to what Jetbrains IDE's do?

@Snailedlt that layout option is featured in the walkthrough video above if I'm not mistaken—added an example below. 

Given that this model accounts for different layout modes, I'm not sure how the Jetbrains example would work on something like the mixed layout (incoming and current in two columns on the top row, result in one row below) that we've generally been discussing here. Let me know if you have any ideas on how that might work.

<img width="1272" alt="a" src="https://user-images.githubusercontent.com/25163139/161676451-92803065-c63f-4b10-9604-0aff83d78c3e.png">

**Comment by abhijitchikane** [comment]
> well in vertical layout I think we can just show the 
![image](https://user-images.githubusercontent.com/50770619/161677810-c833dae8-736f-4165-80a8-215d1278dc9a.png)

this arrow 
I am not designer though Just saying

**Comment by Snailedlt** [comment]
> > well in vertical layout I think we can just show the 
> ![image](https://user-images.githubusercontent.com/50770619/161677810-c833dae8-736f-4165-80a8-215d1278dc9a.png)
> 
> this arrow 
> 
> 
> and if we implement the Jet brains like view I don't think we will ever think of using any other layout

Agreed.
For an even more generic icon, we can just use a checkmark for accepting the changes, and an x for declining them.

So, same concept, just different icons

**Comment by Snailedlt** [comment]
Another thing:
When selecting a change in for example the current pane, I'd like for that change to go unmarked in the current pane. Again, this is like what Jetbrains does. It makes it easier to see which changes you've already handled, and gives a less cluttered view. Also if we could see when all conflicts have been handled that would be great! Maybe by reflecting the state of the "Accept Merge" button (change color and/or text), and also show a popup asking you if you're sure you want to sccept the merge if you've selected none of the changes for a conflict.

This is almost exactly what Jetbrains does, so use that as an interactive example.


Sidenote: the "accept merge" button should be more visible imo, atleast when all the conflicts are solved, so as to indicate that it's the next step in solving the conflict. Right now it's not obvious in which order to press the buttons for absolute beginners.

**Comment by cadsuara** [comment]
This improvement looks fantastic!! However, I think it could be even better. For instance adding support to something like the default layout in the vimdiff git mergetool option. That option shows 'LOCAL', 'BASE', 'REMOTE' and 'MERGED' files
```
vimdiff
			Opens vim with a 4 windows layout distributed in the following way:
 
			    ------------------------------------------
			    |             |           |              |
			    |   LOCAL     |   BASE    |   REMOTE     |
			    |             |           |              |
			    ------------------------------------------
			    |                                        |
			    |                MERGED                  |
			    |                                        |
			    ------------------------------------------
 
			"LOCAL", "BASE" and "REMOTE" are read-only buffers showing the contents of the
			conflicting file in a specific git commit ("commit you are merging into",
			"common ancestor commit" and "commit you are merging from" respectively)

			"MERGED" is a writable buffer where you have to resolve the conflicts (using the
			other read-only buffers as a reference). Once you are done, save and exit "vim"
			as usual (":wq") or, if you want to abort, exit using ":cq".
```
Check also this [blog](https://www.rosipov.com/blog/use-vimdiff-as-git-mergetool/)

An even better implementation that will require more development effort is to follow the same approach as in [this change](https://lore.kernel.org/git/20220330191909.294610-1-greenfoo@u92.eu/) being submitted to the git project. The proposal is to have a variable describing the layout, using some default value that is 'comfortable' for everybody.

**Comment by AndersMad** [comment]
as @cadsuara wrote - here with another example of an excellent compare tool that has 3+1: https://www.scootersoftware.com/features.php - see https://www.scootersoftware.com/images/TextMerge.png

**Comment by jairbubbles** [comment]
Hi @daviddossett, I really like the direction, good job!

I would personally vote for the checkboxes (like in Visual Studio) it's just so easier to understand! In many tools you have custom buttons which are not standard, for instance in Araxis Merge:  
![image](https://user-images.githubusercontent.com/15875066/163339865-6a5ec71f-de2a-46ec-bafe-e35636fd4562.png)
For the "Take All" / "Take Incoming", a checkbox which reflects the global state (all selected / some selected / none selected) could be a nice addition.

Having the branch names to "current" / "incoming" is also super cool but I would like to emphasize that given the context the logic would be inverted. For instance when you do a merge it's quite understandable as "current" is indeed the checked out branch but when you do a rebase it's inverted. "current" will be the branch on which you're rebasing and your checked out branch will be on the right panel. It's very disturbing for people who are not very fluent with git. I believe most users would always their current branch to be on the left.

As for displaying "base" (in a 3 + 1 layout) I don't think it should be the default but there should be a way to display it. I would like to point out how it was resolved in Sublime Merge by allowing to toggle the result view with the base view: 

![image](https://user-images.githubusercontent.com/15875066/163341641-213586d1-ee31-4865-8ad6-802d21d5cb8d.png)

**Comment by Rchatru** [comment]
> I spoke to @isidorn about this offline, but I'd _really_ like for this feature to work for any diff, not just merge conflicts. I think if that was done, it would basically solve #25887 as well. Plus, it would seriously make my day.

I was also glad to see this issue in the roadmap, I think it can be a good way to unblock the famous and claimed #25887, you can't imagine how many people are waiting for this feature.

**Comment by AnrDaemon** [comment]
Checkboxes assume the lines you merge are independent. Which may be true for some formats, but nearly always false for code.
The Meld's way (for example) of expressing operations is way more clear.
Also, read-only source editors are questionable. While most of the time I find them sufficient, for some problematic conflicts I'd like to edit the source to avoid seeing insignificant conflicts and get a clearer view on more important areas.

**Comment by hediet** [comment]
> While most of the time I find them sufficient, for some problematic conflicts I'd like to edit the source to avoid seeing insignificant conflicts and get a clearer view on more important areas.

Do you happen to have a real world example? A screen recording would be amazing!

**Comment by AnrDaemon** [comment]
Consider a class file where you and your colleague added a new method each and modified one of the existing methods.
The new methods are clear additions, and the modified method is a conflicting change.
Merging all changes into common version would leave change marks in 3 places on each source panel. While actually only one change is significant.

**Comment by JuanmaMenendez** [comment]
We need much space as possible, will be better if we open the conflict solving view in a Modal that covers the full screen (or most of it 80%) and set the "column layout" mode by default as this is the mode clearest to understand by far.

The merge conflict window of WebStorm is perfect, we should use it as a  close reference.

**Comment by IllusionMH** [comment]
Floating windows aren't here yet #10121

**Comment by AnrDaemon** [comment]
> merge conflict window of WebStorm

Dare to add a screenshot?

**Comment by isidorn** [comment]
Locking conversation for now. Thank you very much for your feedback. We will keep this issue updated once we have more details.

**Comment by daviddossett** [comment]
Closing now that this is available as an experimental feature via `"git.mergeEditor": true`.

Tracking specific issues with the merge-editor label going forward 👍
